### DISCUSSION TITLE
2021-08-29 (manual pre-release)

### DISCUSSION THREAD

**Discussion by pipcet** [pipcet/pearl#13] [Announcements]
This "release" gets me to a GNOME desktop (Full HD only for now, no audio, initrd) on a Mac Mini connected to the internet via the built-in Ethernet device. Instructions in the discussion.

<hr /><em>This discussion was created from the release <a href='https://github.com/pipcet/pearl/releases/tag/2021-08-29-manual'>2021-08-29 (manual pre-release)</a>.</em>

**Comment by pipcet** [comment]
Very rough instructions for how to do this. Please feel free to contact me or respond here if there are any questions, comments, etc, including if you'd like to test but would like to wait until the process is a bit smoother...

1. download pearl.macho (from this release page)
2. download debian.cpio.zst (from this release page)
3. install pearl.macho as kernel (see other documentation)
4. connect a USB keyboard. Do this before rebooting.
5. boot into pearl
6. in the menu, using the keyboard, select "booting", then "rootshell"
7. wait for a shell prompt
8. enter
```
cd /sys/bus/platform/drivers/pcie-apple
```
9. enter
```
echo 690000000.pcie > bind
```
and keep repeating that command until it succeeds. (There may be a slight delay as loading the xhci driver fails. That's currently normal.)
10. in all likelihood, the ethernet device won't show up right away. Enter
```
echo 690000000.pcie > unbind
echo 690000000.pcie > bind
```
until it does. (Again, there may be a slight delay.)
11. enter
```
cd /sys/bus/platform/drivers/apple-dart
```
12. enter
```
echo 681008000.dart > bind
```
and keep repeating it until the error message is "Device or Resource busy".
13. enter
```
echo 682008000.dart > bind
echo 683008000.dart > bind
```
Ideally, you won't get an SError at this point...
14. enter
```
cd /sys/bus/pci/drivers/tg3
ls
```
and note down the name of the only numeric file in that directory. In my case, it's `0002:03:00.0`, but yours may differ.
15. enter
```
echo 0002:03.00.0 > unbind
echo 0002:03.00.0 > bind
```
replacing, in both lines, `0002:03:00.0` with the name you found out in step 14.
16. use `ifconfig` to set up an Ethernet link. There's now a delay of a minute until the driver decides you're really running it without firmware.
```
ifconfig eth0 192.168.234.2 up
```
replacing `192.168.234.2` by your local IP address
17. use `wget` to download debian.cpio (unpack it first on the remote machine).
```
cd /
wget http://192.168.234.1:8080/debian.cpio
```
replacing `192.168.234.1:8080` with the IP and port name of your HTTP server.
18. unpack `debian.cpio` into a newly-created directory:
```
mkdir /debroot
mount -t ramfs none /debroot
cd /debroot
cpio -id < /debian.cpio
```
19. switch root to the debian rootfs
```
exec switch_root /debroot /bin/bash
```
20. run the second "native" stage of debootstrap
```
/debootstrap/debootstrap --second-stage
```
21. change your root password
```
passwd
X
X
```
22. start debian
```
exec /sbin/init
```
23. log in as root
24. set up a default route for the network which should still exist
```
route add default gw 192.168.234.1
echo nameserver 8.8.8.8 > /etc/resolv.conf
```
replacing `192.168.234.1` by your gateway's IP address
25. do whatever
```
apt-get update
apt-get install task-gnome-desktop
apt-get clean
useradd user
passwd user
init 5
```

**Reply by Johnnie390** [reply]
I would suggest removing points 17 to 20?? in the documentation above as debian.cpio has become superfluous and not everyone has access to their own HTTP server..

Regards

Ry

**Comment by Johnnie390** [comment]
The above documentation is sufficient to arrive at a desktop as described above. The solution is not permanent (setup does not survive reboot) however.

Regards,

Ry

**Comment by pipcet** [comment]
Thank you very much for testing!

You're quite right, the initial Debian system lives only in a ramdisk; sorry, I should have made that clearer.

I'll have to read the debian installer documentation to decide on the best way to launch the debian installer from the ramdisk; something like https://gist.github.com/varqox/42e213b6b2dde2b636ef might currently be required.

The next things to check, if you have the time and motivation, would be whether the /dev/nvme* devices show up and correspond to your disks, and whether 4k works once you run
```
cd /usr
tar xvf /linux.modules
modprobe apple
```

If the former works, we can install a permanent Debian (or whichever distribution you prefer!) system; if the latter works, it'll actually be usable...

Thanks again!
Pip

**Comment by Johnnie390** [comment]
Pip, NVME block devices (/dev/nvme*) _do appear_ when I issue a lsblk command or similiar.
The 4K part will have to wait until the morrow as it is late afternoon here already.

Regards,

Ry

**Reply by pipcet** [reply]
That's excellent! Thanks again for testing so much!

(I would advise repartitioning the mini's drive in macOS, if you want to install anything on the internal drive. A single partition is enough, you can use LVM2 (and probably want to use cryptsetup) to create logical volumes as you need them. It's probably a good idea to give it a recognizable name/label as you can then ensure you won't overwrite another partition).

**Comment by Johnnie390** [comment]
Update:
Pip, I cannot see/find the linux.modules tar file anywhere... so I cannot test your 4K screen scenario.
 
As stated yesterday, from an LXQT (impermanent-will-not-survive-reboot) Desktop, the _ls_, and _lsblk_ commands show me the NVME block device geometries also the blocking, partition sizes etc. look correct. 

At this stage, a permanent Debian Desktop should be feasible methinks. The Ethernet device(s) bind/rebind dance causes headaches though.

Regards

Ry

PS The Desktop started yesterday with a 1920x1080 resolution which is perfectly OK with me at the moment.

**Reply by pipcet** [reply]
Thanks again! The `linux.modules` file should be in the root directory of the Debian ramdisk. If it's not, something has gone wrong and I'll investigate.

I think it would be nice to run the official Debian installer, but it appears more difficult than I thought at first to tell it not to bother installing a bootloader and kernel...

That said, it should be possible to manually set up a Debian system on the nvme now.

You're right, I should look at the pcie code to figure out why the cumbersome PCI initialization code is currently required.

**Reply by Johnnie390** [reply]
Pipcet, I could not find linux.modules anywhere.

Ry

**Reply by pipcet** [reply]
If that's still the case with the pre-release I just triggered, can you try booting "linux" from the pearl menu rather than waiting for debian to boot from initrd? That should give you a root shell and `linux.modules` is in `/boot` or `/persist` (ideally, both). You can then run
```
mount /dev/nvme0n1p4 /mnt
cp /boot/linux.modules /mnt
exit
```
and the system should continue booting automatically without going through the Debian initramfs installation procedure at all.

**Reply by Johnnie390** [reply]
Pip, perhaps we need to open a new thread here. 
The documentation is all over the place and we are not in sync... :(

Ry.

**Comment by Johnnie390** [comment]
Screenshots of my activities yesterday. 
![WhatsApp Image 2021-09-05 at 11 21 01](https://user-images.githubusercontent.com/22815859/132121920-8417b4cc-63e6-478d-ae8b-7607fc107569.jpeg)
![WhatsApp Image 2021-09-05 at 11 21 32](https://user-images.githubusercontent.com/22815859/132121923-3a8f9430-97e5-4bb6-ad9c-a45f046e1043.jpeg)

A little bit of fine tuning and persistence across reboots et voilà.

Ry

PS This is the desktop using the manual, pre-released pearl.macho and debian.cpio (2021-08-29).

**Reply by pipcet** [reply]
Wow, thank you!

Oh, was that the system where you couldn't find `linux.modules`? I think I added the `cp` only after that...

**Reply by Johnnie390** [reply]
AFAIK both releases with and without the Debian.cpio. You need to update the docs to reflect what release to use and procedures.. confused now!!

I am not getting very far with pearl-debian.macho & co. The release from the 29th of August (with debian.cpio - manual) brings me the furthest.

**Reply by Johnnie390** [reply]
Pip, should I try the release from today? What is the _exact_ procedure please?

Ry

**Reply by pipcet** [reply]
Sorry, was busy writing down the instructions and testing the release. Maybe we can continue the discussion at #18 ?
